experimental/Rings/WeierstrassPreparationTheorem.jl

```diff
@@ -0,0 +1,240 @@
+##############################################################
+#
+#Conversion from Singular (weierstr.lib)
+#
+#############################################################
+
+export weierstrass_preparation_theorem
+
+export weierstrass_division_theorem
+
+
+"""
+`order_of_generality(f::MPolyElem, n::Int)`
+
+compute for a given polynomial `f` an integer `d` such that `f` is `x_n`-general of order `d` 
+with `x_n` the n-th variable of the polynomial ring of `f` 
+"""
+function order_of_generality(f::MPolyElem, n::Int)
+	 
+	vect = fill(parent(f)(0), nvars(parent(f))) 
+	vect[n] = gens(parent(f))[n] 
+	f = evaluate(f, vect) 
+	f != 0 || return false, -1
+	return true, degree(leading_monomial(f, ordering=neglex(gens(parent(f)))), n)
+end
+
+"""
+`jet(f::MPolyElem, degree_f::Int)`
+
+Given a polynomial `f` and an integer degree `degree_f` , this function returns the polynomial,
+which contains only the monomials of `f`, which total degree is lower than or equal to `degree_f`
+"""
+function jet(f::MPolyElem, degree_f::Int) 
+	while f != 0
+		if total_degree(leading_monomial(f, ordering=deglex(gens(parent(f))))) > degree_f
+			f = f-coeff(f, leading_monomial(f, ordering=deglex(gens(parent(f)))))*leading_monomial(f, ordering=deglex(gens(parent(f))))
```

Comment:
It seems you are computing `leading_monomial(f, ordering=deglex(gens(parent(f))` three times here? Better to store it in a variable... and also the ordering you keep re-creating...

```suggestion
    ordering = deglex(gens(parent(f)))
	while !iszero(f)
	    lm = leading_monomial(f, ordering)
		if total_degree(lm) > degree_f
			f = f - coeff(f,lm)*lm
```

That said, I am pretty sure this code is very inefficient in its current form. What you really want is to iterate over all *terms* and discard those whose total degree is too big. That does not require a specific monomial ordering; in particular the "intrinsic" ordering of the polynomial implementation could be used...

Reply:
My idea was, to use this ordering to not iterate over all terms of f. But maybe this is inefficient. So i changed it now to 
  i = length(f)
  while i >= 1
    if total_degree(term(f,i)) > degree_f
      f = f - term(f,i)
    end
    i = i - 1
  end

I hope, this matches your idea.